core/TypePtr.cc

```diff
@@ -112,6 +114,10 @@ int TypePtr::kind() const {
             return 11;
         case Tag::SelfType:
             return 12;
+        case Tag::LiteralIntegerType:
+            return 13;
+        case Tag::FloatLiteralType:
+            return 14;
```

Comment:
Do you mind if we renumber the tags to keep all the `LiteralType` things together?

I was looking into some generics bugs a while back, and one of the fixes I had planned was based on assuming that the `TypeVar` and `LambdaParam` things come between the ground types and the type constructors like And and Or.

Also selfishly I prefer `T.any(Integer, String)` to `T.any(String, Integer)` (an ordering we could get back by moving back up).

I _think_ I can write some tests involving generics that would require these literal types to be moved up earlier, but I haven't really spent time on that branch in a while, and would love not to have to spend to much time thinking of test cases right now.

I can add a note to my other branch ([code here](https://github.com/sorbet/sorbet/compare/jez-no-short-circuit-type-constraints)) to write tests involving all literal types when i get around to reviving it.